src/pint/toa.py

```diff
@@ -1336,34 +1404,187 @@ def __len__(self):
         return len(self.table)
 
     def __getitem__(self, index):
-        if not hasattr(self, "table"):
-            raise ValueError("This TOAs object is incomplete and does not have a table")
-        if isinstance(index, np.ndarray) and index.dtype == bool:
-            r = copy.deepcopy(self)
-            r.table = r.table[index]
-            if len(r.table) > 0:
-                r.table = r.table.group_by("obs")
-            return r
-        elif (
-            isinstance(index, np.ndarray)
-            and index.dtype == np.int64
-            or isinstance(index, list)
-        ):
-            r = copy.deepcopy(self)
-            r.table = r.table[index]
-            if len(r.table) > 0:
-                r.table = r.table.group_by("obs")
-            return r
-        elif isinstance(index, slice):
-            r = copy.deepcopy(self)
-            r.table = r.table[index]
-            if len(r.table) > 0:
-                r.table = r.table.group_by("obs")
-            return r
-        elif isinstance(index, int):
-            raise ValueError("TOAs do not support extraction of TOA objects (yet?)")
+        """Extract a subset of TOAs and/or a column/flag from each one.
+
+        When selecting a column from ``self.table`` or a flag from ``self.table["flags"]``,
+        pass its name as a string (not including the ``-`` if it's a flag). The result will
+        be a :class:`~astropy.table.Column` if the string is the name of a column in ``self.table``
+        or a newly allocated array of strings if the string is the name of a flag. If the
+        flag is not set for some or all of the TOAs, this array will contain the empty string
+        at the corresponding place.
+
+        When selecting a subset of the TOAs, a list/array of indices will
+        result in selecting those TOAs (though not necessarily in that order),
+        a list/array of Booleans will result in selecting those TOAs for which
+        the list/array has True, and a slice will select the corresponding
+        slice of TOAs (again, the required grouping by observatory may mean
+        that reordering TOAs will not work as expected).
+
+        Both a column and subset can be selected at once by forming a tuple, as in
+        ``toas["fish", ::10]``; this will result in selecting that subset of the
+        appropriate column. This mode will allow selection of individual elements
+        (``toas["fish", 17]``), unlike all-column selection.
+
+        Parameter
+        ---------
+        index : str or pair or list or array or slice
+            How to choose what to select.
+
+        Returns
+        -------
+        pint.toa.TOAs or np.ndarray or astropy.table.Column
+            The selected part of the object.
+
+        Note
+        ----
+        This function does not currently support extracting a single :class:`~pint.toa.TOA` object,
+        to use integer indexing a column must be selected.
```

Comment:
Can you give an example of the wrong way and then the right way to do that?

Reply:
Done.